sources/packages/backend/libs/services/src/notifications/notification/notification.service.ts

```diff
@@ -70,57 +80,130 @@ export class NotificationService extends RecordDataModelService<Notification> {
+   * Call GCNotify to send email notification.
+   * @param pollingLimit
+   * @param notificationsProcessed optional param used to
+   * increment count on recursion.
+   * @param notificationsSuccessfullyProcessed optional param used to
+   * increment count on recursion.
+   * @returns processing summary.
    */
-  async sendEmailNotification(
-    notificationId: number,
-    entityManager?: EntityManager,
-  ): Promise<GCNotifyResult> {
-    const notificationRepo =
-      entityManager?.getRepository(Notification) ?? this.repo;
-    const notification = await notificationRepo.findOne({
-      select: {
-        id: true,
-        messagePayload: true,
-      },
-      where: {
-        id: notificationId,
-      },
-    });
-    // Call GC Notify send email method.
-    const gcNotifyResult = await this.gcNotifyService.sendEmailNotification(
-      notification.messagePayload,
-    );
+  async processUnsentNotifications(
+    pollingLimit: number,
+    notificationsProcessed = 0,
+    notificationsSuccessfullyProcessed = 0,
+  ): Promise<NotificationProcessingStatus> {
+    const notificationsToProcess = await this.repo
+      .createQueryBuilder("notification")
+      .select(["notification.id", "notification.messagePayload"])
+      .where("notification.dateSent IS NULL")
+      .limit(pollingLimit)
+      .getMany();
+
+    if (notificationsToProcess.length) {
+      this.logger.log(
+        `Processing ${notificationsToProcess.length} notifications`,
+      );
+
+      // Used to limit the number of asynchronous operations
+      // that will start at the same time.
+      const maxPromisesAllowed = os.cpus().length;
+
+      // Hold all the promises that must be processed.
+      const promises: Promise<boolean>[] = [];
+
+      for (const notification of notificationsToProcess) {
+        promises.push(this.sendEmailNotification(notification));
 
-    // Update date sent column in notification table after sending email notification successfully.
-    await this.updateNotification(notification.id, entityManager);
+        if (promises.length >= maxPromisesAllowed) {
+          // Waits for all be executed.
+          const processingResult = await Promise.all(promises);
+          notificationsSuccessfullyProcessed =
+            notificationsSuccessfullyProcessed +
+            processingResult.filter((result) => result).length;
+          // Clear the array.
+          promises.splice(0, promises.length);
+        }
+      }
 
-    return gcNotifyResult;
+      if (promises.length > 0) {
+        // Waits for methods, if any outside the loop.
+        const processingResult = await Promise.all(promises);
+        notificationsSuccessfullyProcessed =
+          notificationsSuccessfullyProcessed +
+          processingResult.filter((result) => result).length;
+      }
+      //Assign the value for total notifications processed.
+      notificationsProcessed =
+        notificationsProcessed + notificationsToProcess.length;
```

Comment:
Do we need to accumulate the values at this point or just after the recursive call on line 160?